File: GFG POTD/020-July-2025/029-ascii-range-sum.cpp
// https://www.geeksforgeeks.org/problems/ascii-range-sum/1




#include <bits/stdc++.h>
using namespace std;

class Solution {
  public:
    // TC : O(N)
    // SC : O(1)
    vector<int> asciirange(string& s) {
        int n = s.length();
        vector<pair<int, int>> firstOcc(26, {-1, -1}), lastOcc(26, {-1, -1});
        
        int asciiSum = 0;
        vector<int> uniques;
        for (int i = 0; i < n; i ++) {
            int idx = s[i] - 'a';
            asciiSum += s[i];
            if (firstOcc[idx].first == -1) {
                firstOcc[idx].first  = i;
                firstOcc[idx].second = asciiSum;
                uniques.push_back(idx);
            } else {
                lastOcc[idx].first  = i;
                lastOcc[idx].second = asciiSum;
            }
        }

        vector<int> result;
        for (int i = 0; i < uniques.size(); i++) {
            int idx =  uniques[i];
            int firstOccIdx = firstOcc[idx].first,  lastOccIdx = lastOcc[idx].first;
            int firstOccSum = firstOcc[idx].second, lastOccSum = lastOcc[idx].second;
            
            int resultantSum = lastOccSum - firstOccSum - (idx + 'a');
            if (resultantSum > 0) result.push_back(resultantSum);
        }
        return result;
    }
};

File: GFG POTD/020-July-2025/029-ascii-range-sum.java
// https://www.geeksforgeeks.org/problems/ascii-range-sum/1




import java.util.*;

class Solution {
    public ArrayList<Integer> asciirange(String s) {
        int n = s.length();
        int[][] firstOcc = new int[26][2];
        int[][] lastOcc  = new int[26][2];
        
        for (int[] row : firstOcc) Arrays.fill(row, -1);
        for (int[] row : lastOcc) Arrays.fill(row, -1);
        
        int asciiSum = 0;
        List<Integer> uniques = new ArrayList<>();
        
        for (int i = 0; i < n; i ++) {
            int idx = s.charAt(i) - 'a';
            asciiSum += s.charAt(i);
            
            if (firstOcc[idx][0] == -1) {
                firstOcc[idx][0] = i;
                firstOcc[idx][1] = asciiSum;
                uniques.add(idx);
            } else {
                lastOcc[idx][0] = i;
                lastOcc[idx][1] = asciiSum;
            }
        }
        
        ArrayList<Integer> result = new ArrayList<>();
        for (int idx : uniques) {
            int resultantSum = lastOcc[idx][1] - firstOcc[idx][1] - (idx + 'a');
            if (resultantSum > 0) {
                result.add(resultantSum);
            }
        }
        return result;
    }
}